fuzz/fuzz_packfile_raw.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * libgit2 raw packfile fuzz target.
+ *
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include <openssl/sha.h>
+
+#include "git2.h"
+
+#define REPOSITORY_PATH "fuzz.git"
+
+static git_repository *repo = NULL;
+static git_odb *odb = NULL;
+
+int LLVMFuzzerInitialize(__attribute__((unused)) int *argc,
+			 __attribute__((unused)) char ***argv)
+{
+	if (git_libgit2_init() < 0) {
+		fprintf(stderr, "Failed to initialize libgit2\n");
+		abort();
+	}
+	if (git_repository_init(&repo, REPOSITORY_PATH, true) < 0) {
+		fprintf(stderr, "Failed to initialize the repository\n");
+		abort();
+	}
+	if (git_repository_odb(&odb, repo) < 0) {
+		fprintf(stderr, "Failed to create the odb\n");
+		abort();
+	}
+	return 0;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
+{
+	if (size == 0)
+		return 0;
+
+	if (!repo || !odb) {
+		fprintf(stderr, "Global state not initialized\n");
+		abort();
+	}
+
+	git_indexer *indexer;
+	git_transfer_progress stats = {0, 0};
+	if (git_indexer_new(&indexer, REPOSITORY_PATH "/objects/pack", 0, odb,
+			    NULL, NULL) < 0) {
+		fprintf(stderr, "Failed to create the indexer: %s\n",
+			giterr_last()->message);
+		abort();
+	}
+
+	/*
+	 * If the first byte in the stream has the high bit set, append the
+	 * SHA1 hash so that the packfile is somewhat valid.
+	 */
+	bool append_hash = *data & 0x80;
+	++data;
+	--size;
+
+	if (git_indexer_append(indexer, data, size, &stats) < 0)
+		goto cleanup;
+	if (append_hash) {
+		char hash[20] = {0};
+		if (SHA1(data, size, (unsigned char *)hash) == NULL) {
+			fprintf(stderr, "Failed to compute the SHA1 hash\n");
+			abort();
+		}
+		if (git_indexer_append(indexer, hash, sizeof(hash), &stats) <
+		    0) {
+			goto cleanup;
+		}
+	}
+	if (git_indexer_commit(indexer, &stats) < 0)
+		goto cleanup;
+
+cleanup:
+	git_indexer_free(indexer);
+	return 0;
```

Comment:
How is the return value being interpreted by LLVM?

Reply:
It is required to be zero (for now): http://llvm.org/docs/LibFuzzer.html#fuzz-target